src/sqlfluff/linter.py

```diff
@@ -838,37 +839,68 @@ def lint_string(self, s, fname="<string input>", fix=False, config=None):
         return linted_file
 
     def paths_from_path(
-        self, path, ignore_file_name=".sqlfluffignore", ignore_non_existent_files=False
+        self,
+        path,
+        ignore_file_name=".sqlfluffignore",
+        ignore_non_existent_files=False,
+        ignore_files=True,
+        working_path=os.getcwd(),
     ):
         """Return a set of sql file paths from a potentially more ambigious path string.
 
         Here we also deal with the .sqlfluffignore file if present.
 
+        When a path to a file to be linted is explicitly passed
+        we look for ignore files in all directories that are parents of the file,
+        up to the current directory.
+
+        If the current directory is not a parent of the file we only
+        look for an ignore file in the direct parent of the file.
+
         """
         if not os.path.exists(path):
             if ignore_non_existent_files:
                 return []
             else:
                 raise IOError("Specified path does not exist")
 
-        # Files referred to exactly are never ignored.
-        if not os.path.isdir(path):
-            return [path]
+        # Files referred to exactly are also ignored if
+        # matched, but we warn the users when that happens
+        is_exact_file = not os.path.isdir(path)
+
+        if is_exact_file:
+            dirpath = os.path.dirname(path)
+            files = [os.path.basename(path)]
+            loader = ConfigLoader.get_global()
+            ignore_file_paths = loader.find_ignore_config_files(
+                path=path, working_path=working_path, ignore_file_name=ignore_file_name
+            )
```

Comment:
see comment above about instantiating the `ConfigLoader`. I think it's wierd that we have to here, but given your other changes I think we can just make it a classmethod and then never need to call `get_global()`.